gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -231,13 +326,136 @@ private ManagedChannel getNextChannel() {
    *     reverse is not true: Two calls with different affinities might return the same channel.
    *     However, the implementation should attempt to spread load evenly.
    */
-  ManagedChannel getChannel(int affinity) {
-    int index = affinity % channels.size();
+  private Entry getEntry(int affinity) {
+    List<Entry> localEntries = entries.get();
+
+    int index = affinity % localEntries.size();
     index = Math.abs(index);
     // If index is the most negative int, abs(index) is still negative.
     if (index < 0) {
       index = 0;
     }
-    return channels.get(index);
+
+    return localEntries.get(index);
+  }
+
+  /** Bundles a gRPC {@link ManagedChannel} with some usage accounting. */
+  private static class Entry {
+    private final ManagedChannel channel;
+    private final AtomicInteger outstandingRpcs = new AtomicInteger(0);
+
+    // Flag that the channel should be closed once all of the outstanding RPC complete.
+    private final AtomicBoolean shutdownRequested = new AtomicBoolean();
+    // Flag that the channel has been closed.
+    private final AtomicBoolean shutdownInitiated = new AtomicBoolean();
+
+    private Entry(ManagedChannel channel) {
+      this.channel = channel;
+    }
+
+    /**
+     * Try to increment the outstanding RPC count. The method will return false if the channel is
+     * closing and the caller should pick a different channel. If the method returned true, the
+     * channel has been successfully retained and it is the responsibility of the caller to release
+     * it.
+     */
+    private boolean retain() {
+      // register desire to start RPC
+      outstandingRpcs.incrementAndGet();
+
+      // abort if the channel is closing
+      if (shutdownRequested.get()) {
+        release();
+        return false;
+      }
+      return true;
```

Comment:
What if you reached this point (`shutdownRequested.get()` returns false above), but another thread requested shutdown (concurrently) right after the check? The method will return true while the Entry got the shutdown request? Is that OK?

---

Another example: the Javadoc says "If the method returned true, the channel has been successfully retained." But let's consider this scenario:

Thread B: calls `retain()` and enjoys using it.

Some time later...

Thread B: calls `release()`
Thread B: runs `outstandingRpcs.decrementAndGet()`, which returns 0. (The local `newCount` is 0.)
Thread A: calls `retain()`
Thread A: runs `outstandingRpcs.incrementAndGet()`. The count is now 1. `retain()` exits with true.
Thread C: calls `requestShutdown()`. `shutdownRequested` is set to true. However, the count is currently 1, so it doesn't call `shutdown()`. The method exits. So far, so good.
Thread B: checks `newCount == 0 && shutdownRequested.get()`, which is true. It calls `shutdown()`.

So, what happens is that, the caller of `retain()` (Thread A) got true, but the channel is already closed.